Project 1: Data cleaning 
Loading the data set 

In [1]:
import pandas as pd
df = pd.read_csv(r"C:\Users\user\Downloads\healthcare messy dataset.csv")

Checking for the info of the dataset to know the detailed information of the data set.

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Patient_ID           10000 non-null  object 
 1   Age                  10000 non-null  int64  
 2   Gender               9370 non-null   object 
 3   BMI                  9236 non-null   float64
 4   Admission_Date       9600 non-null   object 
 5   Admission_Type       9723 non-null   object 
 6   Diagnosis            9351 non-null   object 
 7   Systolic_BP          9273 non-null   float64
 8   Diastolic_BP         9271 non-null   float64
 9   Glucose_Level        9078 non-null   float64
 10  Cholesterol          9110 non-null   float64
 11  Hemoglobin           9261 non-null   float64
 12  Medication           8426 non-null   object 
 13  Smoking_Status       9304 non-null   object 
 14  Insurance_Type       9663 non-null   object 
 15  Length_of_Stay_Days  9435 non-null   

Previewing the first few rows of the dataset.

In [3]:
df.head()

,Patient_ID,Age,Gender,BMI,Admission_Date,Admission_Type,Diagnosis,Systolic_BP,Diastolic_BP,Glucose_Level,Cholesterol,Hemoglobin,Medication,Smoking_Status,Insurance_Type,Length_of_Stay_Days,Doctor_ID,Hospital,Readmission,Outcome
0,P13751,60,Male,25.9,NaN,Urgent,NaN,126.0,73.0,NaN,165.9,12.8,NaN,Non-Smoker,NaN,4.0,DR537,Metro Health Center,No,Improved
1,P12962,49,Male,NaN,02/14/2020,Emergency,Diabetes Type 2,NaN,69.0,77.0,212.0,11.7,Metformin,Never,Medicare,2.0,DR409,Metro Health Center,No,Improved
2,P83901,63,Male,31.9,2023-01-27,NaN,Diabetes Type 2,102.0,85.0,68.7,211.6,14.4,Metformin,Smoker,Private,-1.0,DR993,city general hospital,No,Discharged
3,P91953,79,Male,19.1,14-09-2019,Emergency,Coronary Artery Disease,116.0,74.0,66.6,NaN,14.7,Metoprolol,Non-Smoker,Medicaid,4.0,DR666,Metro Health Center,0,Discharged
4,P47430,47,Female,26.8,09/01/2023,Urgent,Coronary Artery Disease,NaN,NaN,103.4,187.7,13.3,Metoprolol,Smoker,Private,8.0,DR942,St. Mary Medical,No,TRANSFERRED


Checking for missing values and identifying columns with missing data.

In [4]:
df.isnull().sum()

Patient_ID                0
Age                       0
Gender                  630
BMI                     764
Admission_Date          400
Admission_Type          277
Diagnosis               649
Systolic_BP             727
Diastolic_BP            729
Glucose_Level           922
Cholesterol             890
Hemoglobin              739
Medication             1574
Smoking_Status          696
Insurance_Type          337
Length_of_Stay_Days     565
Doctor_ID               268
Hospital                710
Readmission             486
Outcome                 545
dtype: int64

Checking the number of rows and columns in the dataset.

In [5]:
df.shape

(10000, 20)

Creating a copy of the dataset to preserve the original data.

In [6]:
df.copy() 
healthcare = df.copy()

Identifying and counting duplicate records in the dataset.

In [7]:
healthcare.duplicated(subset = 'Patient_ID').sum()

695

Removing duplicate records while retaining the first occurence of each.

In [8]:
healthcare = healthcare.drop_duplicates(subset = 'Patient_ID' , keep='first')

Comfirming that all duplicate records have been successfully removed.

In [9]:
healthcare.duplicated(subset='Patient_ID').sum()

0

Checking the format and values of the Admission Date column.

In [10]:
healthcare['Admission_Date'].value_counts().head(10)

Admission_Date
06/10/2021    6
03/07/2022    6
07-01-2024    6
04/14/2020    6
10-08-2018    6
2023-07-23    6
09/23/2023    6
09/17/2023    6
23-06-2020    5
2018-10-16    5
Name: count, dtype: int64

Converting all Admission Date values to a uniform YYYY-MM-DD format.

In [11]:
from datetime import datetime 
def fix_date(val):
    for fmt in ("%Y-%m-%d","%m/%d/%Y","%d-%m-%Y"):
        try:
            return datetime.strptime(str(val),fmt).strftime("%Y-%m-%d")
        except:
            continue
    return None
healthcare['Admission_Date'] = healthcare['Admission_Date'].apply(fix_date)

Confirming all dates are now in the correct YYYY-MM-DD format.

In [12]:
healthcare['Admission_Date'].value_counts().head(10)

Admission_Date
2018-08-10    11
2023-12-02    10
2022-01-04    10
2018-02-18    10
2020-04-14    10
2023-07-23    10
2020-02-02    10
2024-04-03     9
2021-07-24     9
2023-09-26     9
Name: count, dtype: int64

Checking the unique values and distribution of the Gender column.

In [13]:
healthcare['Gender'].value_counts()

Gender
Female    3278
Male      3154
M          778
F          730
male       395
female     382
Name: count, dtype: int64

Replacing inconsistent Gender entries with uniform Male/Female values.

In [14]:
healthcare['Gender'] = healthcare['Gender'].replace({
    'Female': 'Female',
    'female': 'Female',
    'F':'Female',
    'Male':'Male',
    'male':'Male',
    'M':'Male'})

Confirming all Gender values are now consistent.

In [15]:
healthcare['Gender'].value_counts()

Gender
Female    4390
Male      4327
Name: count, dtype: int64

Veryfying the data types of all columns after cleaning.

In [16]:
healthcare.dtypes

Patient_ID              object
Age                      int64
Gender                  object
BMI                    float64
Admission_Date          object
Admission_Type          object
Diagnosis               object
Systolic_BP            float64
Diastolic_BP           float64
Glucose_Level          float64
Cholesterol            float64
Hemoglobin             float64
Medication              object
Smoking_Status          object
Insurance_Type          object
Length_of_Stay_Days    float64
Doctor_ID               object
Hospital                object
Readmission             object
Outcome                 object
dtype: object

Converting the Admission Date column to the correct datetype(datetime).

In [17]:
healthcare['Admission_Date'] = pd.to_datetime(healthcare['Admission_Date'])

Changing Admission Type data type to category.

In [18]:
healthcare['Admission_Type'] = healthcare['Admission_Type'].astype('category')

Checking for inconsistent values in the Smoking Status column.

In [19]:
healthcare['Smoking_Status'].value_counts()

Smoking_Status
Non-Smoker       3251
Former Smoker    2341
Smoker           2098
Never             603
non-smoker        135
smoker            109
former smoker     107
never              25
Name: count, dtype: int64

Replacing inconsistent Smoking Status entries with uniform values.

In [20]:
healthcare['Smoking_Status'] = healthcare['Smoking_Status'].replace({
    'Non-Smoker':'Non-Smoker',
    'Former Smoker':'Former-Smoker',
    'Smoker':'Smoker',
    'Never':'Never',
    'non-smoker':'Non-Smoker',
    'Smoker':'Smoker',
    'former smoker':'Former-Smoker',
    'never':'Never',
    'smoker':'Smoker'})

Standardizing text case in Diagnosis and Admission_Type columns

In [21]:
healthcare['Diagnosis'] = healthcare['Diagnosis'].str.strip().str.title()
healthcare['Admission_Type'] = healthcare['Admission_Type'].str.strip().str.title()

Confirming all Smoking Status values are now consistent.

In [22]:
healthcare['Smoking_Status'].value_counts()

Smoking_Status
Non-Smoker       3386
Former-Smoker    2448
Smoker           2207
Never             628
Name: count, dtype: int64

Filling missing Length of Stay values using the median.

In [23]:
healthcare['Length_of_Stay_Days'] = healthcare['Length_of_Stay_Days'].fillna(healthcare['Length_of_Stay_Days']).median()

Filling missing Doctor ID values with Unknown

In [24]:
healthcare['Doctor_ID'] = healthcare['Doctor_ID'].fillna('Unknown')

Converting negative values to positive using abs() and ensuring correct data type with astype().

In [25]:
healthcare['Length_of_Stay_Days'] = healthcare['Length_of_Stay_Days'].abs().astype(int)

Reviewing the cleaned dataset to confirm all changes were successfully applied.

In [26]:
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9305 entries, 0 to 9997
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Patient_ID           9305 non-null   object        
 1   Age                  9305 non-null   int64         
 2   Gender               8717 non-null   object        
 3   BMI                  8587 non-null   float64       
 4   Admission_Date       8931 non-null   datetime64[ns]
 5   Admission_Type       9047 non-null   object        
 6   Diagnosis            8709 non-null   object        
 7   Systolic_BP          8630 non-null   float64       
 8   Diastolic_BP         8629 non-null   float64       
 9   Glucose_Level        8454 non-null   float64       
 10  Cholesterol          8478 non-null   float64       
 11  Hemoglobin           8616 non-null   float64       
 12  Medication           7840 non-null   object        
 13  Smoking_Status       8669 non-null   o

Filling missing categorical columns with mode and numerical columns with median.

In [27]:
cat_cols = ['Gender','Diagnosis','Medication',
            'Smoking_Status','Insurance_Type',
            'Hospital','Readmission','Outcome']
for col in cat_cols:
    healthcare[col] = healthcare[col].fillna(healthcare[col].mode()[0]).astype('category')

In [28]:
num_cols = ['BMI','Systolic_BP','Diastolic_BP',
            'Glucose_Level','Cholesterol','Hemoglobin']
for col in num_cols:
    healthcare[col] = healthcare[col].fillna(healthcare[col].median())

Confirming all missing values have been successfully handled.

In [29]:
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9305 entries, 0 to 9997
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Patient_ID           9305 non-null   object        
 1   Age                  9305 non-null   int64         
 2   Gender               9305 non-null   category      
 3   BMI                  9305 non-null   float64       
 4   Admission_Date       8931 non-null   datetime64[ns]
 5   Admission_Type       9047 non-null   object        
 6   Diagnosis            9305 non-null   category      
 7   Systolic_BP          9305 non-null   float64       
 8   Diastolic_BP         9305 non-null   float64       
 9   Glucose_Level        9305 non-null   float64       
 10  Cholesterol          9305 non-null   float64       
 11  Hemoglobin           9305 non-null   float64       
 12  Medication           9305 non-null   category      
 13  Smoking_Status       9305 non-null   c

Filling missing Admission Type values with the most frequent value(mode).

In [30]:
healthcare['Admission_Type'] = healthcare['Admission_Type'].fillna(healthcare['Admission_Type'].mode()[0])

Filling missing Doctor ID values with the most frequent value(mode).

In [31]:
healthcare['Doctor_ID'].fillna(healthcare['Doctor_ID'].mode()[0])

0       DR537
1       DR409
2       DR993
3       DR666
4       DR942
        ...  
9993    DR439
9994    DR841
9995    DR792
9996    DR945
9997    DR947
Name: Doctor_ID, Length: 9305, dtype: object

Filling missing Admission Date values with the most frequent value(mode).

In [32]:
healthcare['Admission_Date'] = healthcare['Admission_Date'].fillna(healthcare['Admission_Date'].mode()[0])

Confirming all columns now have 9305 non-null vahlues with no missing data remaining.

In [33]:
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9305 entries, 0 to 9997
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Patient_ID           9305 non-null   object        
 1   Age                  9305 non-null   int64         
 2   Gender               9305 non-null   category      
 3   BMI                  9305 non-null   float64       
 4   Admission_Date       9305 non-null   datetime64[ns]
 5   Admission_Type       9305 non-null   object        
 6   Diagnosis            9305 non-null   category      
 7   Systolic_BP          9305 non-null   float64       
 8   Diastolic_BP         9305 non-null   float64       
 9   Glucose_Level        9305 non-null   float64       
 10  Cholesterol          9305 non-null   float64       
 11  Hemoglobin           9305 non-null   float64       
 12  Medication           9305 non-null   category      
 13  Smoking_Status       9305 non-null   c

Confirming zero missing values across all columns.

In [34]:
healthcare.isnull().sum()

Patient_ID             0
Age                    0
Gender                 0
BMI                    0
Admission_Date         0
Admission_Type         0
Diagnosis              0
Systolic_BP            0
Diastolic_BP           0
Glucose_Level          0
Cholesterol            0
Hemoglobin             0
Medication             0
Smoking_Status         0
Insurance_Type         0
Length_of_Stay_Days    0
Doctor_ID              0
Hospital               0
Readmission            0
Outcome                0
dtype: int64

Confirming zero deplicate records remain in the cleaned data set.

In [35]:
healthcare.duplicated().sum()

0

Confirming all columns have the correct data types after cleaning.

In [36]:
healthcare.dtypes

Patient_ID                     object
Age                             int64
Gender                       category
BMI                           float64
Admission_Date         datetime64[ns]
Admission_Type                 object
Diagnosis                    category
Systolic_BP                   float64
Diastolic_BP                  float64
Glucose_Level                 float64
Cholesterol                   float64
Hemoglobin                    float64
Medication                   category
Smoking_Status               category
Insurance_Type               category
Length_of_Stay_Days             int32
Doctor_ID                      object
Hospital                     category
Readmission                  category
Outcome                      category
dtype: object

Reviewing the descriptive statistics of all numerical columns after cleaning.

In [37]:
healthcare.describe()

,Age,BMI,Admission_Date,Systolic_BP,Diastolic_BP,Glucose_Level,Cholesterol,Hemoglobin,Length_of_Stay_Days
count,9305.000000,9305.000000,9305,9305.000000,9305.000000,9305.000000,9305.000000,9305.000000,9305.0
mean,51.465019,28.509952,2021-05-31 03:52:17.281031680,124.150242,76.363568,106.634788,178.934573,13.491725,3.0
min,1.000000,12.000000,2018-01-01 00:00:00,-1.000000,31.000000,8.300000,65.600000,6.400000,3.0
25%,39.000000,24.800000,2019-07-20 00:00:00,111.000000,70.000000,91.500000,157.200000,12.400000,3.0
50%,51.000000,28.500000,2021-05-29 00:00:00,120.000000,76.000000,106.300000,179.000000,13.500000,3.0
75%,64.000000,32.200000,2023-03-18 00:00:00,130.000000,83.000000,121.600000,200.300000,14.600000,3.0
max,100.000000,53.100000,2024-12-31 00:00:00,999.000000,110.000000,211.300000,309.000000,20.200000,3.0
std,17.947577,5.728255,NaN,65.209681,9.668089,24.717042,34.355314,1.738943,0.0


Verifying there are no negative Systolic BP values or extreme Glucose Level values above 500.

In [38]:
print(healthcare[healthcare['Systolic_BP'] < 0].shape[0])
print(healthcare[healthcare['Glucose_Level'] > 500].shape[0])

49
0


Replacing negative, zero and unrealistic values(999) in Systolic BP with the median to ensure data integrity.

In [39]:
median_sbp = healthcare[healthcare['Systolic_BP'] > 0]['Systolic_BP'].median()
healthcare.loc[healthcare['Systolic_BP'] < 0 ,'Systolic_BP'] = median_sbp
healthcare.loc[healthcare['Systolic_BP'] == 0,'Systolic_BP'] = median_sbp
healthcare.loc[healthcare['Systolic_BP'] == 999,'Systolic_BP'] = median_sbp

Confirming minimum and maximum Systolic BP values are now within realistic range.

In [40]:
print('Min Systolic BP:', healthcare['Systolic_BP'].min())
print('Max Systolic BP:', healthcare['Systolic_BP'].max())

Min Systolic BP: 55.0
Max Systolic BP: 250.0


Confirming all numerical columns now have realistic values.

In [41]:
healthcare.describe()

,Age,BMI,Admission_Date,Systolic_BP,Diastolic_BP,Glucose_Level,Cholesterol,Hemoglobin,Length_of_Stay_Days
count,9305.000000,9305.000000,9305,9305.000000,9305.000000,9305.000000,9305.000000,9305.000000,9305.0
mean,51.465019,28.509952,2021-05-31 03:52:17.281031680,121.022354,76.363568,106.634788,178.934573,13.491725,3.0
min,1.000000,12.000000,2018-01-01 00:00:00,55.000000,31.000000,8.300000,65.600000,6.400000,3.0
25%,39.000000,24.800000,2019-07-20 00:00:00,111.000000,70.000000,91.500000,157.200000,12.400000,3.0
50%,51.000000,28.500000,2021-05-29 00:00:00,120.000000,76.000000,106.300000,179.000000,13.500000,3.0
75%,64.000000,32.200000,2023-03-18 00:00:00,129.000000,83.000000,121.600000,200.300000,14.600000,3.0
max,100.000000,53.100000,2024-12-31 00:00:00,250.000000,110.000000,211.300000,309.000000,20.200000,3.0
std,17.947577,5.728255,NaN,17.369695,9.668089,24.717042,34.355314,1.738943,0.0


Exporting the final cleaned dataset to CSV for use in project 2.

In [42]:
healthcare.to_csv('healthcare_cleaned.csv', index=False)
print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


Conclusion
The dataset has been successfully cleaned.